packages/light-client-extension-helpers/src/background/json-rpc-provider.ts

```diff
@@ -0,0 +1,205 @@
+import type * as smoldot from "../smoldot"
+
+import type {
+  JsonRpcConnection,
+  JsonRpcProvider,
+} from "@polkadot-api/json-rpc-provider"
+import { getSyncProvider } from "@polkadot-api/json-rpc-provider-proxy"
+import {
+  Cause,
+  Effect,
+  Fiber,
+  pipe as $,
+  Runtime,
+  Schedule,
+  SynchronizedRef,
+  Queue,
+  Stream,
+  Sink,
+  PubSub,
+} from "effect"
+
+export type Client = Pick<smoldot.Client, "addChain">
+export type AddChainOptions = smoldot.AddChainOptions
+
+/**
+ * Creates a new JSON RPC provider using a client and options.
+ *
+ * This provider will automatically re-connect to the chain if the
+ * connection is lost.
+ *
+ * @param client - The client to use for creating the provider.
+ * @param options - The options for adding a chain.
+ */
+export const make = (
+  client: Client,
+  options: AddChainOptions,
+): Effect.Effect<JsonRpcProvider, Cause.UnknownException, never> => {
+  return Effect.gen(function* () {
+    const runtime = yield* Effect.runtime<never>()
+    const runCallback = Runtime.runCallback(runtime)
+    const runPromise = Runtime.runPromise(runtime)
+
+    /**
+     * This is a special single-element queue that signals when the daemon is
+     * ready. It is cleared every time the daemon restarts. While a `Deferred`
+     *  would typically be used in this situation, the ability of the daemon
+     * to restart requires a queue that can be cleared and reused.
+     */
+    const daemonReadySignal = yield* Queue.sliding<void>(1)
+
+    /**
+     * A message pubsub for the daemon to publish JSON RPC messages to.
+     * It is the responsibility of the JSON RPC provider to consume these
+     * messages and propagate them with the onMessage callback.
+     */
+    const messagePubSub = yield* PubSub.unbounded<string>()
+
+    /**
+     * An error pubsub for the daemon to publish errors to.
+     * It is the responsibility of the JSON RPC provider to consume these
+     *  errors and propagate them with the onError callback.
+     */
+    const errorPubSub = yield* PubSub.sliding<void>(1)
+
+    /**
+     * A synchronized reference to the current chain. This reference is updated
+     * every time the daemon restarts.
+     *
+     */
+    const chainRef = yield* $(
+      Effect.tryPromise(() => client.addChain(options)),
+      Effect.andThen(SynchronizedRef.make),
+    )
+
+    /**
+     * This daemon process listens for JSON RPC messages, publishing them to
+     * the message pubsub. It also detects errors and publishes those to the
+     * error pubsub.
+     *
+     * If an error occurs with `chain.nextJsonRpcResponse()`, the daemon
+     * restarts. This indicates that the chain is no longer functional.
+     * In response, the daemon reconnects to smoldot using the same
+     * `AddChainParameters` to obtain a new `Chain` object.
+     */
+    const daemon = yield* Effect.gen(function* () {
```

Comment:
I wonder if breaking a couple of bits into separate functions with suitable names would help eg

```ts
function streamMessagesToPubSub(client, options, messagePubSub): Fiber.RuntimeFiber<void, Cause.UnknownException> {

}
```

(I think that's what this does right? It just calls addChain and then streams `chain.nextJsonRpcResponse()` to a message queue, restarting the whole thing in case of error)